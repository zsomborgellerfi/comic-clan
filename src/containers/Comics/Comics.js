import React from "react";
import { connect } from "react-redux";
import Loading from "../../components/Loading";
import ComicsList from "../../components/ComicsList";
import { LinkContainer } from "react-router-bootstrap";

import { getComics } from "../../actions";
import "./Comics.scss";
import { comicFilterKeys } from "../../utils/comicFilterKeys";
import { Form, InputGroup, Button, Container } from "react-bootstrap";
import { IoIosSearch } from "react-icons/io";
import _ from "lodash";

class Comics extends React.Component {
  constructor(props) {
    super(props);
    const urlCategory = props.match.params.category;
    this.state = {
      activeFilterKey:
        urlCategory && comicFilterKeys.includes(urlCategory)
          ? urlCategory
          : comicFilterKeys[0],
      comicsGroupedByKey: {},
      searchText: "",
    };
  }

  componentDidMount() {
    this.fetchComics();
  }

  componentDidUpdate(prevProps, prevState) {
    const { comics, location, match } = this.props;
    if (prevState.searchText !== this.state.searchText) {
      this.fetchComics();
    }
    if (
      comics.length !== prevProps.comics.length ||
      location.pathname !== prevProps.location.pathname
    ) {
      this.filterComics(match.params.category || this.state.activeFilterKey);
    }
  }

  fetchComics = () => {
    this.props.getComics(this.state.searchText);
  };

  filterComics = (key) => {
    const { comics } = this.props;

    //group comics by selected filter key
    const groupBy = (key) => (array) =>
      array.reduce(
        (objectsByKeyValue, obj) => ({
          ...objectsByKeyValue,
          [obj[key]]: (objectsByKeyValue[obj[key]] || []).concat(obj),
        }),
        {}
      );

    let comicsGroupedByKey = {};

    if (key === "random") {
      comicsGroupedByKey = { [key]: _.shuffle(comics).slice(0, 5) };
    } else {
      comicsGroupedByKey = groupBy(key)(comics);
    }

    this.setState({ activeFilterKey: key, comicsGroupedByKey });
  };

  onSearchInputChange = (event) => {
    /* signal to React not to nullify the event object */
    event.persist();

    if (!this.debouncedFn) {
      // debounce fetch requests while typing
      this.debouncedFn = _.debounce(() => {
        let searchText = event.target.value;
        this.setState({ searchText });
      }, 300);
    }

    this.debouncedFn();
  };

  render() {
    const { loading } = this.props;
    const { comicsGroupedByKey, activeFilterKey } = this.state;

    return (
      <Container fluid className={"Comics"}>
        <div>
          <Form onSubmit={(e) => e.preventDefault()}>
            <InputGroup className="Comics__searchInputWrapper">
              <InputGroup.Prepend>
                <InputGroup.Text className={"Comics__searchIcon"}>
                  <IoIosSearch size={24} color={"#cccccc"} />
                </InputGroup.Text>
              </InputGroup.Prepend>
              <Form.Control
                placeholder="Search by book name"
                onChange={this.onSearchInputChange}
                className={"Comics__searchInput"}
              />
            </InputGroup>
          </Form>
          {comicFilterKeys.map((key) => {
            return (
              <LinkContainer
                key={key}
                to={`/comics/${key}`}
                className={`Comics__filterSwitch ${
                  activeFilterKey === key ? "Comics__filterSwitchActive" : ""
                }`}
              >
                <Button>{key}</Button>
              </LinkContainer>
            );
          })}
          {!loading ? (
            Object.keys(comicsGroupedByKey).map((key, idx) => {
              return (
                <div key={key}>
                  {idx > 0 && <hr className={"separator"} />}
                  <h3 className={"Comics__rowTitle"}>{key}</h3>
                  <ComicsList comics={comicsGroupedByKey[key]} />
                </div>
              );
            })
          ) : (
            <Loading />
          )}
        </div>
      </Container>
    );
  }
}

Comics.defaultProps = {
  comics: [],
};

const mapStateToProps = (state) => {
  return {
    comics: state.comics,
    loading: state.loading,
  };
};
const mapDispatchToProps = {
  getComics: getComics,
};

export default connect(mapStateToProps, mapDispatchToProps)(Comics);
