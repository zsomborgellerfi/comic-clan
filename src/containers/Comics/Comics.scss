@import '../../styles/variables';

.Comics {
  background: $bg-color-2;
  min-height: $full-height;
  padding-top: 30px;
  &__rowTitle {
    margin: 30px 0;
    font-size: 32px;
    color: $text-color-5;
    font-weight: normal;
    text-transform: capitalize;
  }
  &__searchInputWrapper {
    background-color: transparent;
    border: 3px solid #777777;
    border-radius: 8px;
    padding: 10px 18px;
    margin-bottom: 30px;
  }
  &__searchIcon {
    background-color: transparent;
    border: none;
  }
  &__searchInput {
    background-color: transparent;
    border: none;
    padding: 0;
    font-weight: 500;
    font-size: 20px;
    line-height: 23px;
    color: $text-color-1;

    &:focus {
      background-color: transparent;
      border: none;
      outline: none;
      color: $text-color-1;
    }
  }
  &__filterSwitch {
    margin-right: 15px;
    border-radius: 20px;
    background-color: transparent;
    border: none;
    outline: none;
    text-transform: capitalize;
    color: $text-color-4;
    font-weight: bold;
    font-size: 18px;
    line-height: 21px;
    padding: 10px 20px;
    &Active,
    &:active,
    &:focus,
    &:hover {
      background-color: $button-color !important;
      border-color: $button-color !important;
      box-shadow: none !important;
      color: white;
    }
    
  }
  .separator {
    margin: 70px 0;
  }
}
