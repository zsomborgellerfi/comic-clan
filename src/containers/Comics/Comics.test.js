import React from "react";
import { shallow } from "enzyme";
import { Comics } from "./Comics";

const props = {
  getComics: jest.fn(),
  comics: [{ name: "test-name", rating: 3 }],
  match: {
    params: {
      category: "year",
    },
  },
};
describe("ComicDetails", () => {
  const wrapper = shallow(<Comics {...props} />);

  it("should render Comics correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
