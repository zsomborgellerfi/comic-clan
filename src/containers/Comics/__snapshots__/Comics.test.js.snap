// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComicDetails should render Comics correctly 1`] = `
<Container
  className="Comics"
  fluid={true}
>
  <div>
    <Form
      inline={false}
      onSubmit={[Function]}
    >
      <InputGroup
        className="Comics__searchInputWrapper"
      >
        <InputGroupPrepend>
          <InputGroupText
            className="Comics__searchIcon"
          >
            <IoIosSearch
              color="#cccccc"
              size={24}
            />
          </InputGroupText>
        </InputGroupPrepend>
        <FormControl
          className="Comics__searchInput"
          onChange={[Function]}
          placeholder="Search by book name"
        />
      </InputGroup>
    </Form>
    <withRouter(LinkContainer)
      className="Comics__filterSwitch Comics__filterSwitchActive"
      key="year"
      to="/comics/year"
    >
      <Button
        active={false}
        disabled={false}
        type="button"
        variant="primary"
      >
        year
      </Button>
    </withRouter(LinkContainer)>
    <withRouter(LinkContainer)
      className="Comics__filterSwitch "
      key="writer"
      to="/comics/writer"
    >
      <Button
        active={false}
        disabled={false}
        type="button"
        variant="primary"
      >
        writer
      </Button>
    </withRouter(LinkContainer)>
    <withRouter(LinkContainer)
      className="Comics__filterSwitch "
      key="artist"
      to="/comics/artist"
    >
      <Button
        active={false}
        disabled={false}
        type="button"
        variant="primary"
      >
        artist
      </Button>
    </withRouter(LinkContainer)>
    <withRouter(LinkContainer)
      className="Comics__filterSwitch "
      key="owner"
      to="/comics/owner"
    >
      <Button
        active={false}
        disabled={false}
        type="button"
        variant="primary"
      >
        owner
      </Button>
    </withRouter(LinkContainer)>
    <withRouter(LinkContainer)
      className="Comics__filterSwitch "
      key="random"
      to="/comics/random"
    >
      <Button
        active={false}
        disabled={false}
        type="button"
        variant="primary"
      >
        random
      </Button>
    </withRouter(LinkContainer)>
  </div>
</Container>
`;
